### ISSUE TITLE
automatic folds from for loops

### ISSUE THREAD

**Issue by johnynek** [johnynek/bosatsu#20]
a common idiom in python is something like:
```python
x = 0
for y in z:
  x += y
return x
```
This does not look so functional, but it could be. We could convert for comprehensions like this into folds. We could do this by considering `x += y` to be rebinding a new `x to x + y` and we could look at all such rebound variables in the loop. Then we could convert it to something like:
```
z.foldLeft(0, \y, x -> x + y)
```
This is not so dissimilar to having the `do` notation in haskell which allows you to write imperative looking code that still returns a value.

Similarly, we could do the same translation for map-like for comprehensions:
`[f(x) for x in y]` to `y.map(f)`

There may be some subtlety with making these fully composable which we would want to support.

**Comment by johnynek** [comment]
list comprehensions were done in #174 

Still have not thought more about for loops that rebind variables. I think they can all be rewritten as folds on lists.

**Comment by johnynek** [comment]
actually, I think a better way to think about this is that a `for` loop is a let binding + a fold. It is always setting up a let binding for any values that are assigned in the for loop:
```
sum = 0
for x in xs:
  sum = sum + x
fn(sum)
```
this is the same as:
```
sum = xs.fold_left(sum, \sum, x -> sum + x)
fn(sum)
```
or:
```
sum = 0
count = 0
for x in xs:
  count = count + 1
  sum = sum + x

fn(count, sum)
```
becomes:
```
(count, sum) = xs.fold_left((0, 0), \(count, sum), x -> (count + 1, sum + x))
fn(count, sum)
```

I think this translation is workable and would make it more convenient for people not expert in writing loops as folds.

**Comment by avibryant** [comment]
I think you mean `sum = xs.fold_left(0, \sum, x -> sum + x)`.

This is great. The only thing I worry about is that the initialization of eg `sum` here is part of the same expression as the `for` but that's not obvious; what happens if you insert some unrelated expression (like a println) in between them?

**Comment by johnynek** [comment]
so, I think you rewrite the body of the for loop so you have an expression that returns a tuple of all the bindings. Since all the code is pure, that rewrite is safe. So, you can just replace the final expression in the for loop with `(count, sum)` and it should be correct I think.

**Comment by avibryant** [comment]
let me give a more concrete example. Is something like this legal?

```
sum = 0
z = sum*2
for x in xs:
  sum = sum + x
fn(sum, z)
```

**Comment by johnynek** [comment]
yeah, that's fine. translate to (noting rebinding lets is already allowed):
```
sum = 3
z = sum * 2
sum = xs.foldLeft(sum, \sum, x ->
  sum = sum + x
  sum)
fn(sum, z)
```
so, you just build a tuple of the values bound in the for loop, then put those bindings into a tuple and return it a fold.
Isn't that what you intend?

**Comment by johnynek** [comment]
Okay, I took a stab at this and it's quite a bit more complex than I had originally thought.

Consider the following:

```python
even_sum = 0
for x in xs:
  if is_even(x):
    even_sum = add(even_sum, x)
  #implicitly there is an else: even_sum = even_sum here
```

Similarly, we should allow trailing bindings inside a match inside a for:

```python
even_sum = 0
for x in xs:
  match is_even(x):
    case True:
       even_sum = add(even_sum, x)
    case False:
       pass # in the context of a binding for this is the same setting all the binding values to themselves.
```

(I think match should still be total since we want to train people to think about totality).

Perhaps the right design is to make a new sealed trait that represents BindingSyntax. We want a BindingIf to be able to have other BindingSyntax in the branches. Also we want a BindingMatch to have BindingSyntax in the case bodies. We want a BindingFor to have BindingSyntax in the loop body. So, this should allow nested for loops. We should probably have `pass` which is kind of an empty BindingSyntax. Of course, the normal pattern bind would be the kind of basic BindindSyntax. Then the Declaration be something like:
```scala
// this is only a sketch, we need region/padding/optindent stuff, etc...
case class Binding(bind: BindingSyntax, rest: Declaration) extends Declaration
```
If we had that, we could use it in match branches to represent assigning a branch to itself.

Of course, this will make rewriting Binding into Expr considerably harder: both the for loop and the match can introduce new variable which may shadow the outer binding variables. Secondly, if there are repeated fors, you will have to do
nested folds, which will need to be converted inside to the outside, but since the inner loops may be assigning to outer loop variables, you will need to be careful to make sure to carefully compute the set of bindings for the inner loops and track which they bind.

I think it is tractable to solve, but the compilation is much harder. Next, what I worry about is that type-checking happens at the Expr layer, but the syntax is converted before we do typing. So the checks we can do at the syntax layer are more restricted, so the errors at type checking may be very confusing. We may need to solve #962 first so that we can report type errors better.
